src/replication/replication_node_manager.cpp

```diff
@@ -61,44 +75,78 @@ namespace cubreplication
 
     void finalize ()
     {
+      wait_ha_tasks ();
+
       g_hostname.clear ();
 
       delete g_slave_node;
       g_slave_node = NULL;
       delete g_master_node;
       g_master_node = NULL;
 
-      // stream and stream file are interdependent, therefore first stop the stream
+      // we need to first stop the stream before destroying stream_file
       g_stream->stop ();
       delete g_stream_file;
       g_stream_file = NULL;
       delete g_stream;
       g_stream = NULL;
     }
 
-    void commute_to_master_state ()
+    void commute_to_master_state (cubthread::entry *thread_p, bool force)
     {
-      delete g_slave_node;
-      g_slave_node = NULL;
-
-      if (g_master_node == NULL)
-	{
-	  g_master_node = new master_node (g_hostname.c_str (), g_stream, g_stream_file);
-	}
+      wait_ha_tasks ();
+      inc_tasks ();
+
+      cubthread::entry_task *promote_task = new cubthread::entry_callable_task ([thread_p, force] (
+		  cubthread::entry &context)
+      {
+	if (g_slave_node != NULL)
+	  {
+	    g_slave_node->wait_fetch_completed ();
+	  }
+	delete g_slave_node;
+	g_slave_node = NULL;
+
+	if (g_master_node == NULL)
+	  {
+	    g_master_node = new master_node (g_hostname.c_str (), g_stream, g_stream_file);
+	  }
+
+	css_finish_transit (thread_p, force, HA_SERVER_STATE_ACTIVE);
+	dec_tasks ();
+	g_commute_cv.notify_all ();
+      }, true);
+
+      auto wp = cubthread::internal_tasks_worker_pool::get_instance ();
+      cubthread::get_manager ()->push_task (wp, promote_task);
     }
 
-    void commute_to_slave_state ()
+    void commute_to_slave_state (cubthread::entry *thread_p, bool force)
     {
-      // todo: remove after master -> slave transitions is properly handled
-      assert (g_master_node == NULL);
-
-      delete g_master_node;
-      g_master_node = NULL;
-
-      if (g_slave_node == NULL)
-	{
-	  g_slave_node = new cubreplication::slave_node (g_hostname.c_str (), g_stream, g_stream_file);
-	}
+      wait_ha_tasks ();
+      inc_tasks ();
```

Comment:
rename as inc_ha_tasks ; it should be clear is the same sort of 'task' we are waiting for in the prev line of code